### PR TITLE
improve type annotations in 'docutils.statemachine'

### PR DESCRIPTION
**Pull request by danieleades**
adds type annotations for docutils.statemachine.

I'm not super confident about the use of the generic `_Context` for the state machine since the only code examples I can find using this class are in the `sphinx` repo, and there no context is used (ever).

### CODE REVIEW

**Comment by danieleades** [comment]
@srittau I could really use your eyes on this one! I think i've made good progress, but could use some assistance to get this over the line, if you have the time

**Comment by danieleades** [comment]
i tried to simplify some of these type signatures using type aliases, but couldn't get this to work

```python
_Context = TypeVar("_Context")
_TransitionResult: TypeAlias = tuple[_Context, str | None, list[str]]
_TransitionMethod: TypeAlias = Callable[[Match[str], _Context, str], _TransitionResult[_Context]]
```

mypy and pyright complain about the generics being unbound

**Comment by JelleZijlstra** [comment]
> i tried to simplify some of these type signatures using type aliases, but couldn't get this to work
> 
> ```python
> _Context = TypeVar("_Context")
> _TransitionResult: TypeAlias = tuple[_Context, str | None, list[str]]
> _TransitionMethod: TypeAlias = Callable[[Match[str], _Context, str], _TransitionResult[_Context]]
> ```
> 
> mypy and pyright complain about the generics being unbound

That looks right, but you should use the alias as `_TransitionResult[_Context]` in this case.

**Comment by danieleades** [comment]
> That looks right, but you should use the alias as `_TransitionResult[_Context]` in this case.

i've reinstated the aliases to show what i mean. Neither mypy nor pyright is happy about them

**Comment by JelleZijlstra** [comment]
```
stubs/docutils/docutils/statemachine.pyi:3: error: Module "typing" has no attribute "TypeAlias"  [attr-defined]
stubs/docutils/docutils/statemachine.pyi:3: note: Use `from typing_extensions import TypeAlias` instead
```

This probably has something to do with the other errors.

**Comment by danieleades** [comment]
@JelleZijlstra anything else needed to get this one merged? (it's blocking a couple of other PRs)

**Comment by danieleades** [comment]
Note i've requested reviews from some Sphinx maintainers since this 'context' generic type will propagate through a great many docutils types and have a significant impact on downstream users of these stubs

**Comment by JelleZijlstra** [comment]
I haven't had a chance to look at this in much detail. The generics are weird to me because each class just seems to use the TypeVar as a type argument to other generics, until you get to `_TransitionResult`. You say that:

> I'm not super confident about the use of the generic `_Context` for the state machine since the only code examples I can find using this class are in the `sphinx` repo, and there no context is used (ever).

Assuming that Sphinx is a typical user of docutils, that might mean making these classes generic is going to be more of an annoyance than a help to many users. Therefore maybe we shouldn't be making these generic, at least until we're able to use PEP-696. (Which might not be far away, as the PEP has been accepted and the upcoming mypy release adds some support.)

**Comment by danieleades** [comment]
> I haven't had a chance to look at this in much detail. The generics are weird to me because each class just seems to use the TypeVar as a type argument to other generics, until you get to `_TransitionResult`. You say that:
> 
> > I'm not super confident about the use of the generic `_Context` for the state machine since the only code examples I can find using this class are in the `sphinx` repo, and there no context is used (ever).
> 
> Assuming that Sphinx is a typical user of docutils, that might mean making these classes generic is going to be more of an annoyance than a help to many users. Therefore maybe we shouldn't be making these generic, at least until we're able to use PEP-696. (Which might not be far away, as the PEP has been accepted and the upcoming mypy release adds some support.)

I was looking at this for this PR actually.

Thing is, since i've not been able to find any uses of the 'context', i'm not sure what the default type should be...

**Comment by JelleZijlstra** [comment]
> Thing is, since i've not been able to find any uses of the 'context', i'm not sure what the default type should be...

The default could be `object`, so users who don't use it just get an opaque object.

**Comment by danieleades** [comment]
> > Thing is, since i've not been able to find any uses of the 'context', i'm not sure what the default type should be...
> 
> The default could be `object`, so users who don't use it just get an opaque object.

i've elected to remove the generics for now to unblock this, and will revisit in a separate PR

**Comment by danieleades** [comment]
replaced `Any` with `list[str]` for context based on this comment - https://github.com/sphinx-doc/sphinx/issues/12007#issuecomment-1975377720

File: stubs/docutils/docutils/statemachine.pyi
Comment (picnixz) [inline stubs/docutils/docutils/statemachine.pyi:16]:
```suggestion
    current_state: str
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (picnixz) [inline stubs/docutils/docutils/statemachine.pyi:6]: I think it should be fine to have an alias for the StateName so that we know that `str | None` is a state name or None.

File: stubs/docutils/docutils/statemachine.pyi
Comment (picnixz) [inline stubs/docutils/docutils/statemachine.pyi:39]:
```suggestion
    def get_source_and_line(self, lineno: int | None = None) -> tuple[str, int] | tuple[None, None]: ...
```

**Comment by danieleades** [comment]
> I'll review more when I'm again in another train but here are some comments based on the docutils HEAD implementation. Can you also check whether they are consistent with the stable version?
> 
> Btw, sorry for the order of my comments but I tend to review things in a random order, and comment when I have an idea (I would have thought that gh would have ordered the comments from top to bottom but it appears not).

Thank you, I appreciate the detailed review. Will go through it when I get time

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi:6]: do you mean this?

```python
_StateName: TypeAlias = str | None
_TransitionResult: TypeAlias = tuple[list[str], _StateName, list[str]]
_TransitionMethod: TypeAlias = Callable[[Match[str], list[str], str], _TransitionResult]
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi:18]: oh confusing.

so based on your description it's not actually `Callable[[str | None, int | None], None]`

it's

`Callable[[str, int], None]` or
`Callable[[str, None], None]` or
`Callable[[None, None], None]` but *not*
`Callable[[None, int], None]`

@JelleZijlstra is there a compact way to annotate that?

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi:47]: makes sense. I'll wait for some agreement on https://github.com/python/typeshed/pull/11469/files#r1511283606

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi:122]: > I think you should use __other so that it indicates (without enforcing) that you expect positional only arguments. I don't know whether you can use positional-only arguments syntax or not.

i haven't noticed that pattern elsewhere in typeshed. I'd like to follow convention here. @JelleZijlstra?

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi:128]: this is the implementation:

```python
    def __lt__(self, other): return self.data < self.__cast(other)   # noqa
    def __le__(self, other): return self.data <= self.__cast(other)  # noqa
    def __eq__(self, other): return self.data == self.__cast(other)  # noqa
    def __ne__(self, other): return self.data != self.__cast(other)  # noqa
    def __gt__(self, other): return self.data > self.__cast(other)   # noqa
    def __ge__(self, other): return self.data >= self.__cast(other)  # noqa
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi:141]: certainly appears to:

```python
    def __mul__(self, n):
        return self.__class__(self.data * n, items=(self.items * n))
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi:164]: see https://github.com/python/mypy/issues/2220

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi]: done

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi]: done

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi]: done

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi]: done

File: stubs/docutils/docutils/statemachine.pyi
Comment (picnixz) [reply stubs/docutils/docutils/statemachine.pyi:6]: Ah yes, it might be redundant but it might help people I think. At least you know what this str is

File: stubs/docutils/docutils/statemachine.pyi
Comment (picnixz) [reply stubs/docutils/docutils/statemachine.pyi:128]: IIRC the `__cast` function is the identity on non-ViewList objects and returns other.data otherwise I think. 

So the expected argument is either a ViewList or a list of data (I think).

File: stubs/docutils/docutils/statemachine.pyi
Comment (picnixz) [reply stubs/docutils/docutils/statemachine.pyi:18]: Actually your callable should support those kind of inputs except for the combination (None, int), but I don't think we can type it like that unless we have an intersection type. So we could technically allow `Callable[[str | None, int | None], None]` even though it would never be called with (None, int) :/

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi]: good catch

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi:109]: feels weird to inherit from a base class that adds mixin methods in a typestub

File: stubs/docutils/docutils/statemachine.pyi
Comment (picnixz) [reply stubs/docutils/docutils/statemachine.pyi:18]: It's actually incorrect because it means that a function that only accept None and None is allowed but then it cannot process the normal inputs being str and int.

**Comment by danieleades** [comment]
@srittau have you got any bandwidth to help me get this one over the line? A few of the other PRs are dependent on this one.

the state machine classes actually appear to be generic over the 'context' parameter, however i've only been able to find subclasses in actual use where the context is `list[str]`, so making the base class generic is possibly of limited value. Personally i'm actually in favour of making the base class generic but removing the generic is probably the more pragmatic solution. as @JelleZijlstra pointed out, a default generic would be a perfect solution.

File: stubs/docutils/docutils/statemachine.pyi
Comment (JelleZijlstra) [reply stubs/docutils/docutils/statemachine.pyi:122]: We should use PEP 570 pos-only syntax in new code, not double underscores. (We had some technical reasons why we didn't migrate all existing code yet.) Here I would make the parameter positional-only if it's also positional-only at runtime. If the class is implemented in Python, it probably isn't.

File: stubs/docutils/docutils/statemachine.pyi
Comment (JelleZijlstra) [reply stubs/docutils/docutils/statemachine.pyi:109]: No need to add the inheritance unless the runtime also inherits from these classes.

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:19]:
```suggestion
    def __init__(self, state_classes: Iterable[type[State]], initial_state: str, debug: bool = False) -> None: ...
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:23]:
```suggestion
        input_lines: Sequence[str] | StringList,
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:25]: It says in the docstring that context is a "application-specific storage", and I only see it passed to states. Would it make more sense to use `Any` or even a type alias `_Context: TypeAlias = Any`?

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:44]:
```suggestion
    def add_states(self, state_classes: Iterable[type[State]]) -> None: ...
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:55]: Considering that these are always filled with not-None values in `__init__`, I'd skip the `| None` part. Might need a stubtest override.

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:53]: According to the docstring:

```suggestion
    patterns: ClassVar[dict[str, str | Pattern[str]] | None]
    initial_transitions: ClassVar[Sequence[str] | Sequence[tuple[str, str]] | None]
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:64]:
```suggestion
    def add_transitions(self, names: Iterable[str], transitions) -> None: ...
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:57]: Considering that a transition is a fairly complex type and it's used multiple times in this class, I'd consider creating a type alias.

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:65]: Shouldn't the transition type match the others in this class?

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:85]: Shouldn't the first return type in all of these be consistently either `list[str]` or `StringList`?

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:91]: As above:

```suggestion
    indent_sm: type[StateMachine]
    indent_sm_kwargs: dict[str, Any]
    known_indent_sm: type[StateMachine]
    known_indent_sm_kwargs: dict[str, Any]
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:92]:
```suggestion
    ws_patterns: dict[str, Pattern[str]]
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:107]: Nit: I'd use `_T` as name, similar to most collection types in typeshed are typed. An move the type var to the top of the file, seeing as how generic it is.

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:116]:
```suggestion
        initlist: ViewList[_Data] | Sequence[_Data] | None = None,
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:128]: Either:

```suggestion
    def __contains__(self, item: _Data) -> bool: ...
```

or

```suggestion
    def __contains__(self, item: object) -> bool: ...
```

(I think that's what @picnixz meant.)

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:131]:
```suggestion
    def __getitem__(self, i: slice) -> Self[_Data]: ...
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:133]: Per @picnixz:

```suggestion
    def __getitem__(self, i: SupportsIndex) -> _Data: ...
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:138]:
```suggestion
    def __setitem__(self, i: SupportsIndex, item: _Data) -> None: ...
    def __delitem__(self, i: SupportsIndex) -> None: ...
```

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:144]:
```suggestion
    def __add__(self, other: ViewList[_Data]) -> Self[_Data]: ...
    def __radd__(self, other: ViewList[_Data]) -> Self[_Data]: ...
    def __iadd__(self, other: ViewList[_Data]) -> Self[_Data]: ...
    def __mul__(self, n: int) -> Self[_Data]: ...
    __rmul__ = __mul__
    def __imul__(self, n: int) -> Self[_Data]: ...
```

**Review by srittau — CHANGES_REQUESTED** [review]
Thanks, remarks below.

File: stubs/docutils/docutils/statemachine.pyi
Comment (AlexWaygood) [reply stubs/docutils/docutils/statemachine.pyi:144]: `Self` can't be parameterised

Comment (srittau) [reply stubs/docutils/docutils/statemachine.pyi:144]: Good point, I fixed it in the comment.

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi:91]: Seems reasonable, though there's a possibility it might be set to None in a subclass and not initialised in __init__

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi]: i did have it set up this way at one point. see https://github.com/python/typeshed/pull/11469#issuecomment-1975938313 and https://github.com/python/typeshed/pull/11469#issuecomment-1986788144

i could reinstate the generics

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [reply stubs/docutils/docutils/statemachine.pyi:91]: I don't know the intricacies of docutils. If that's a possibility, we should keep it as is.

File: stubs/docutils/docutils/statemachine.pyi
Comment (danieleades) [reply stubs/docutils/docutils/statemachine.pyi:61]: it isn't used the same way twice at the moment...

`initial_transitions: ClassVar[Sequence[str] | Sequence[tuple[str, str]] | None]`
`transitions: dict[str, tuple[Pattern[str], Callable[[], None], str]]`
`make_transition(...) -> tuple[Pattern[str], _TransitionMethod[_Context], str]`

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [reply stubs/docutils/docutils/statemachine.pyi]: Ok, sounds good then. You might still consider a type var (to `list[str]`) for documentation purposes, but I leave this decision to you.

File: stubs/docutils/docutils/statemachine.pyi
Comment (srittau) [inline stubs/docutils/docutils/statemachine.pyi:133]: `item` should either be `_T` or `object`. The implementation does just `item in self.data`, where `self.data` is `list[_T]`.

**Review by srittau — CHANGES_REQUESTED** [review]
I spotted one thing. But looks good to go apart from that.

**Review by srittau — APPROVED** [review]

**Pull request merged**
